### PR TITLE
ci: Use `dart pub publish` command for release workflows

### PR DESCRIPTION
**Pull request by Nidal-Bakir**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/Parse-SDK-Flutter/issues?q=is%3Aissue).

### Issue Description

Closes: #881

### Approach

Using the recommended solution by the Flutter team [publishing packages using GitHub actions](https://dart.dev/tools/pub/automated-publishing#publishing-packages-using-github-actions)

### TODOs before merging

@mtrezza There are some steps you need to do in the pub.dev admin tab to enable this.

On the Dart SDK admin tab add the following 

 

On the Flutter SDK admin tab add the following 
 

 

Make sure to follow this [hardening security with tag-protection rules on github](https://dart.dev/tools/pub/automated-publishing#hardening-security-with-tag-protection-rules-on-github)

- [ ] Add tests
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)
- [ ] A changelog entry

### CODE REVIEW

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Review by mtrezza — COMMENTED** [review]
Could you add some checks before publishing? Like we have in the CI where we do lint and analyzer checks before the dry-run.

File: .github/workflows/release-automated.yml [hunk 1]
```diff
@@ -1,29 +1,27 @@
...
+      - "dart-[0-9]+.[0-9]+.[0-9]+*"
+      - "flutter-[0-9]+.[0-9]+.[0-9]+*"
 jobs:
   release:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout code
-        uses: actions/checkout@v3
-      - name: Publish dart package
+        uses: actions/checkout@v3.5.2
```
[the remark this answers is not in the document]
Comment (Nidal-Bakir) [reply .github/workflows/release-automated.yml:13]: done

File: .github/workflows/release-automated.yml [hunk 2]
```diff
@@ -1,29 +1,27 @@
...
+      - name: Setup Dart SDK
+        uses: dart-lang/setup-dart@v1.5.0
```
[the remark this answers is not in the document]
Comment (Nidal-Bakir) [reply .github/workflows/release-automated.yml:15]: done

File: .github/workflows/release-automated.yml [hunk 3]
```diff
@@ -1,29 +1,27 @@
...
-          accessToken: ${{ secrets.PUBDEV_GOOGLE_ACCOUNT_ACCESS_TOKEN  }}
-          refreshToken: ${{ secrets.PUBDEV_GOOGLE_ACCOUNT_REFRESH_TOKEN }}
-          relativePath: packages/dart
-          format: true
-          dryRunOnly: false
-      - name: Publish flutter package
+        run: |
+          cd packages/dart
+          dart pub get
+          dart pub publish --force
```
[the remark this answers is not in the document]
Comment (Nidal-Bakir) [reply .github/workflows/release-automated.yml:21]: force [docs](https://dart.dev/tools/pub/cmd/pub-lish#--force-or--f)

>--force or -f
> With this, pub does not ask for confirmation before publishing. Normally, it shows you the package contents and asks for you to confirm the upload.

**Comment by mtrezza** [comment]
Looks good; could you also modify the manual release workflow? We need to trigger the manual release for the unpublished flutter package.

**Comment by Nidal-Bakir** [comment]
Remember to enable the automated publishing in pub.dev admin tab for each package?

File: .github/workflows/release-automated.yml [hunk 4]
```diff
@@ -1,29 +1,36 @@
...
+  push:
+    tags:
+      - "dart-[0-9]+.[0-9]+.[0-9]+*"
+      - "flutter-[0-9]+.[0-9]+.[0-9]+*"
 jobs:
   release:
     runs-on: ubuntu-latest
+    env:
+     dir: ${{ startsWith(github.ref_name, 'flutter') and 'packages/flutter' or 'packages/dart' }}
+
```
Comment (mtrezza) [inline .github/workflows/release-automated.yml:13]:
```suggestion
```

File: .github/workflows/release-automated.yml [hunk 5]
```diff
@@ -1,29 +1,36 @@
...
+        run: dart analyze --fatal-infos ${{ env.dir }}
+
+      - name: Check lints
+        run: dart format --output=none --set-exit-if-changed ${{ env.dir }}
+
+      - name: Publish --dry-run
+        working-directory: ${{ env.dir }}
+        run: dart pub publish --dry-run
+
+      - name: Publish package
```
Comment (mtrezza) [inline .github/workflows/release-automated.yml:34]:
```suggestion
      - name: Setup Dart SDK
        uses: dart-lang/setup-dart@v1
      - name: Resolve packages
        run: dart pub get --directory ${{ env.dir }}
      - name: Analyze
        run: dart analyze --fatal-infos ${{ env.dir }}
      - name: Check lint
        run: dart format --output=none --set-exit-if-changed ${{ env.dir }}
      - name: Publish package (dry-run)
        working-directory: ${{ env.dir }}
        run: dart pub publish --dry-run
      - name: Publish package
```

File: .github/workflows/release-manual.yml [hunk 6]
```diff
@@ -21,27 +21,31 @@ on:
 jobs:
   release:
     runs-on: ubuntu-latest
+    env:
+     dir: ${{ github.event.inputs.package == 'flutter' and 'packages/flutter' or 'packages/dart' }}
+
```
Comment (mtrezza) [inline .github/workflows/release-manual.yml:26]:
```suggestion
```

File: .github/workflows/release-manual.yml [hunk 7]
```diff
@@ -21,27 +21,31 @@ on:
...
+        run: dart analyze --fatal-infos ${{ env.dir }}
+
+      - name: Check lints
+        run: dart format --output=none --set-exit-if-changed ${{ env.dir }}
+
+      - name: Publish --dry-run
+        working-directory: ${{ env.dir }}
+        run: dart pub publish --dry-run
+
+      - name: Publish package
```
Comment (mtrezza) [inline .github/workflows/release-manual.yml:49]:
```suggestion
      - name: Setup Dart SDK
        uses: dart-lang/setup-dart@v1
      - name: Resolve packages
        run: dart pub get --directory ${{ env.dir }}
      - name: Analyze
        run: dart analyze --fatal-infos ${{ env.dir }}
      - name: Check lint
        run: dart format --output=none --set-exit-if-changed ${{ env.dir }}
      - name: Publish package (dry-run)
        working-directory: ${{ env.dir }}
        run: dart pub publish --dry-run
      - name: Publish package
```

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: .github/workflows/release-automated.yml [hunk 8]
```diff
@@ -1,29 +1,29 @@
 name: release-automated
+
```
Comment (mtrezza) [inline .github/workflows/release-automated.yml:2]:
```suggestion
```

**Review by mtrezza — APPROVED** [review]
Looks good, is this ready for merge?

**Comment by Nidal-Bakir** [comment]
Yes

**Comment by mtrezza** [comment]
It may be that the manual publish workflow doesn't work anymore, the [docs](https://dart.dev/tools/pub/automated-publishing#publishing-packages-using-github-actions) say:

> The tag pattern is a string that must contain {{version}}. Only GitHub Actions triggered by a push of a tag that matches this tag pattern will be allowed to publish your package.

In that case we'd have to delete and re-push a tag to re-trigger a release. We'll see.

**Comment by mtrezza** [comment]
Failed, see https://github.com/parse-community/Parse-SDK-Flutter/issues/881#issuecomment-1548752838

**Pull request merged**